android: Support decode-to-texture Mode

borongc opened the pull request:
Bring up decode-to-texture mode on android, where SbarboardRenderer gets `SbDecodeTarget` from SbPlayer, and assigns it with `gpu::Mailbox` that is used to bind a SharedImage, so Chromium can get the textures from Starboard.

For Chromium, texture mode works as:
decoder generates a texture --> associate with a unique `gpu::Mailbox` --> Renderer creates a SharedImage with the generated `gpu::Mailbox` --> Renderer puts the SharedImage to the rendering pipeline via Render() --> Chromium gets the texture from the SharedImage with the generated `gpu::Mailbox`.

Specifically, `StarboardCodecImage` is to bind a SharedImage with `AndroidVideoImageBacking` on android, so Chromium rendering pipeline can get the texture from SbPlayer, and render it.

Note this PR also adds `StarboardVideoImageBacking` to allow to build it on Linux, but it is not supported yet.

Issue: 375070492

borongc commented:
Note this PR is based on https://github.com/youtube/cobalt/pull/5113, and will be merged after PR 5113.

yell0wd0g commented:
Since I'll be OOO (and hopefully AFK) for a few days, pinging @sherryzy for future reviews.

borongc commented:
> Since I'll be OOO (and hopefully AFK) for a few days, pinging @sherryzy for future reviews.

Still working on this PR, and this is far from ready for be reviewing.

borongc committed:
[android] Support decode-to-texture mode

Bring up decode-to-texture mode on android, where SbarboardRenderer gets `SbDecodeTarget` from SbPlayer, and assigns it with `gpu::Mailbox` that is used to bind a SharedImage, so Chromium can get the textures from Starboard.

For Chromium, texture mode works as:
decoder generates a texture --> associate with a unique `gpu::Mailbox` --> Renderer creates a SharedImage with the generated `gpu::Mailbox` --> Renderer puts the SharedImage to the rendering pipeline via Render() --> Chromium gets the texture from the SharedImage with the generated `gpu::Mailbox`.

Specifically, `StarboardCodecImage` is to bind a SharedImage with `AndroidVideoImageBacking` on android, so Chromium rendering pipeline can get the texture from SbPlayer, and render it.

Note this PR also adds `StarboardVideoImageBacking` to allow to build it on Linux, but it is not supported yet.

Issue: 375070492

borongc commented:
/gemini review

borongc commented on media/gpu/starboard/starboard_gpu_factory_impl.cc:
```diff
@@ -47,6 +63,90 @@ void StarboardGpuFactoryImpl::Initialize(base::UnguessableToken channel_token,
+    textures.push_back(texture);
+  }
+#if BUILDFLAG(IS_ANDROID)
+  DCHECK_EQ(mailboxes.size(), 1);
+  DCHECK_EQ(textures.size(), 1);
```
Android uses single planar.

borongc commented on media/gpu/android/starboard/starboard_codec_image.cc:
```diff
@@ -0,0 +1,78 @@
+
+void StarboardCodecImage::UpdateAndBindTexImage(GLuint service_id) {
+  AssertAcquiredDrDcLock();
+  // TODO(borongchen): update decode-to-target
```
Removed.

borongc commented on media/mojo/services/starboard/starboard_renderer_wrapper.cc:
```diff
@@ -183,4 +276,66 @@
+    provider->gpu_factory_
+        .AsyncCall(&StarboardGpuFactory::RunSbDecodeTargetFunctionOnGpu)
+        .WithArgs(target_function, target_function_context, &done_event);
+    // TODO(borongchen): Blocking is okay here because XYZ.
+    base::ScopedAllowBaseSyncPrimitives allow_wait;
+    done_event.Wait();
```
Done.

borongc commented on content/renderer/media/media_factory.cc:
```diff
@@ -630,7 +630,9 @@ MediaFactory::CreateRendererFactorySelector(
   media::RendererFactoryTraits renderer_factory_traits;
   GetContentClient()->renderer()->GetStarboardRendererFactoryTraits(&renderer_factory_traits);
-  renderer_factory_traits.max_video_capabilities = max_video_capabilities;
+  // TODO(borongchen): remove this for decode-to-texture mode.
+  // renderer_factory_traits.max_video_capabilities = max_video_capabilities;
+  renderer_factory_traits.max_video_capabilities = "width=1920; height=1080; framerate=15; decode-to-texture=true";
```
This is for testing decode-to-texture mode, will be removed.

sherryzy commented:
I'm still reviewing this PR. To help me understand the overall context, would you mind updating the description to clarify what it accomplishes? Given that it was created some time ago and there are several stacked PRs, this would be greatly appreciated.

borongc commented:
> I'm still reviewing this PR. To help me understand the overall context, would you mind updating the description to clarify what it accomplishes? Given that it was created some time ago and there are several stacked PRs, this would be greatly appreciated.

Updated the PR description.

sherryzy commented on starboard/android/shared/decode_target.cc:
```diff
@@ -26,6 +26,8 @@
 #include "starboard/android/shared/jni_env_ext.h"
 #include "starboard/shared/gles/gl_call.h"
 
+#define GL_TEXTURE_NATIVE_ID_ANGLE 0x3481
```
> Otherwise, add a comment linking to where this value comes from.

nit: I think the comment here is reasonable. It's good to add a comment where this value is from.

borongc replied:
Done.

sherryzy commented:
Overall, the change looks good. 

The build's failure for starboard_renderer_wrapper_unittest should be related to this change. Can you fix that?

Just a suggestion, I feel this PR is still a little bit too large for review, and it touches multiple components from the GPU service layer to the media rendering pipeline. It would be beneficial to further split this into smaller, stacked PRs. A suggested approach is:
1) Introduce the foundational shared image classes (StarboardVideoImageBacking, StarboardCodecImage). 
2) Add the StarboardGpuFactory methods for creating these images. 
3) Integrate these components into the StarboardRendererWrapper and the rest of the media pipeline. 

This will allow for more focused reviews of each logical part of the change. However, splitting it up at this stage would require significant rework, we can proceed with reviewing it as is this time.

borongc commented on media/gpu/starboard/starboard_gpu_factory_impl.cc:
```diff
@@ -47,6 +63,92 @@ void StarboardGpuFactoryImpl::Initialize(base::UnguessableToken channel_token,
+    std::vector<gpu::Mailbox>& mailboxes,
+    std::vector<uint32_t>& texture_service_ids,
+#if BUILDFLAG(IS_ANDROID)
+    scoped_refptr<gpu::RefCountedLock> drdc_lock,
```
This is enabled by `features::NeedThreadSafeAndroidMedia()`, and it's only for Android and MAC.

https://source.chromium.org/chromium/chromium/src/+/main:gpu/config/gpu_finch_features.cc;drc=b8b5768e5d0d1c2a84fe4896eae884d97fd1131e;l=221

Other platforms are not supported.

sherryzy reviewed (approved):
LGTM. There is still one build fails. Just make sure all tests pass before you submit this PR.

borongc commented:
Note that Section 2.3 in [go/starboard-renderer-media-pipeline](http://go/starboard-renderer-media-pipeline) illustrates how a VideoFrame binding with a texture in StarboardRenderer.

kjyoun commented on media/gpu/android/starboard/starboard_codec_image.h:
```diff
@@ -0,0 +1,60 @@
+  ~StarboardCodecImage() override;
+
+ private:
+  scoped_refptr<gpu::gles2::TexturePassthrough> texture_passthrough_;
```
nit: const scoped_refptr<gpu::gles2::TexturePassthrough>..

kjyoun commented on media/mojo/services/starboard/gpu_mojo_media_client_starboard.cc:
```diff
@@ -51,7 +56,18 @@ std::unique_ptr<AudioEncoder> CreatePlatformAudioEncoder(
+  }
+
+  return std::make_unique<StarboardRendererWrapper>(std::move(traits),
+                                                    ref_counted_lock);
```
isn't this more concise?
````
return std::make_unique<StarboardRendererWrapper>(std::move(traits),
     features::NeedThreadSafeAndroidMedia() ? 
       base::MakeRefCounted<gpu::RefCountedLock>() : nullptr);
````

kjyoun commented on media/mojo/services/starboard/gpu_mojo_media_client_starboard.cc:
```diff
@@ -51,7 +56,18 @@ std::unique_ptr<AudioEncoder> CreatePlatformAudioEncoder(
+#if BUILDFLAG(IS_ANDROID)
+  scoped_refptr<gpu::RefCountedLock> ref_counted_lock;
+
+  if (features::NeedThreadSafeAndroidMedia()) {
```
can you add comments about why we create `ref_counted_lock` only when `NeedThreadSafeAndroidMedia` is enabled?

kjyoun commented on media/mojo/services/starboard/starboard_renderer_wrapper.cc:
```diff
@@ -211,4 +302,67 @@ void StarboardRendererWrapper::OnRequestOverlayInfoByStarboard(
+    SbDecodeTargetGraphicsContextProvider* graphics_context_provider,
+    SbDecodeTargetGlesContextRunnerTarget target_function,
+    void* target_function_context) {
+  StarboardRendererWrapper* provider =
```
I think here we can use `auto` w/o losing readability, since `reinterpret_cast<StarboardRendererWrapper*>` immediately follows

kjyoun commented on media/mojo/services/starboard/starboard_renderer_wrapper.cc:
```diff
@@ -211,4 +302,67 @@ void StarboardRendererWrapper::OnRequestOverlayInfoByStarboard(
 #endif  // BUILDFLAG(IS_ANDROID)
 
+SbDecodeTargetGraphicsContextProvider*
+StarboardRendererWrapper::GetSbDecodeTargetGraphicsContextProvider() {
```
StarboardRendererWrapper::GetSbDecodeTargetGraphicsContextProvider() const ?

kjyoun commented on media/mojo/services/starboard/starboard_renderer_wrapper.cc:
```diff
@@ -211,4 +302,67 @@ void StarboardRendererWrapper::OnRequestOverlayInfoByStarboard(
+  }
+}
+
+void StarboardRendererWrapper::GetCurrentDecodeTarget() {
```
this method name is little bit confusing, since usually `Get` method return something

`SetCurrentDecodeTarget` ?

kjyoun commented on media/mojo/services/starboard/starboard_renderer_wrapper.cc:
```diff
@@ -211,4 +302,67 @@ void StarboardRendererWrapper::OnRequestOverlayInfoByStarboard(
+  StarboardRendererWrapper* provider =
+      reinterpret_cast<StarboardRendererWrapper*>(
+          graphics_context_provider->gles_context_runner_context);
+
```
CHECK(provider) ?

kjyoun commented on media/mojo/services/starboard/starboard_renderer_wrapper_unittest.cc:
```diff
@@ -171,8 +197,14 @@ class StarboardRendererWrapperTest : public testing::Test {
         std::move(client_extension_remote), base::NullCallback(),
         AndroidOverlayMojoFactoryCB());
+#if BUILDFLAG(IS_ANDROID)
+    scoped_refptr<gpu::RefCountedLock> ref_counted_lock;
```
?
```
renderer_wrapper_ = std::make_unique<StarboardRendererWrapper>(
        std::move(traits), /*ref_counted_lock=*/nullptr);
```

kjyoun reviewed (commented):
Added styel comments
